draft-ietf-quic-spin-exp.md

```diff
@@ -235,11 +235,19 @@ Implementations SHOULD allow administrators of clients and servers to disable
 the spin bit either globally or on a per-connection basis.
 Even when the spin bit is not disabled by the administrator implementations
 SHOULD disable the spin bit on a randomly chosen
-fraction of connections.  The selection process should be designed such that
-on average the spin bit is disabled for at least 1/8th of the connections.
-
-When the spin bit is disabled, endpoints SHOULD set the spin bit value to zero,
-regardless of the values received from their peer. Addendums or revisions to
+fraction of connections.
+
+The selection process SHOULD be designed such that
+on average the spin bit is disabled for at least one eighth of network paths.
+The random choice SHOULD be dependent
+on the source and destination addresses of the path,
+so that the spin bit is consistently enabled or
+disabled for repeated use of the same path.
+
+When the spin bit is disabled, endpoints SHOULD set the spin bit value to
+a constant value randomly chosen to be 0 or 1,
+regardless of the values received from their peer.  Alternatively, endpoints MAY
+change this value when changing connection ID.  Addendums or revisions to
```

Comment:
I have a number of problems with this text:

1. I don't think we need to specify anything
2. If we are to specify something, I would prefer it to be random on a per-packet basis.
3. The combination of SHOULD/MUST isn't even clear about whether other behaviors are permissible.

Reply:
... but I don't like allowing multiple behaviors, because of finger printing risks. What are yoy trying to achieve with a random spin per packet?

Reply:
Well, as I said, I would prefer not specifying anything. As far as I can tell, there's no valid interop reason to do so. What I'm trying to avoid is having a characteristic "I am not doing spin bit" behavior that people anchor on, and "this is always 1" is totally that. Random is at least somewhat better.

As far as needing a PRNG, there's plenty of randomness floating around in packet processing. Specifically, because the PNE mask generation function generates 128 bits at a time.

Reply:
@EKR Removing text is easy, and I would be willing to do that if there is consensus. But I would like to hear another couple of voices advocating that -- say @martinthomson or @kazuho, who expressed opinions on the subject before.

As for PNRG,  it cannot be the PNE mask for the spin bit. The spin bit is protected by the AEAD checksum and has to be set before AEAD protection. The PNE mask is computed using the AEAD output, and is not available in time for randomizing the spin bit.

Reply:
You can't use the PNE mask for creating random data that is part of the data protected by AEAD tag because the PNE process might be offloaded to a separate process or hardware.

Reply:
@huitema: this text is a PR, so it needs consensus to be added at all. As for the PRNG: you can use data generated from the PNE for a previous packet. As I said, there's plenty of surplus. I don't think that @mikkelfj's point is persuasive. Nobody is telling you how to generate random data. I am merely observing that it's effectively free to do so with most implementations.

Again, I think this document should remain silent on the contents of this bit if you aren't doing spin. My position is: silent > random > the current text.

Reply:
I am concerned with fingerprinting. If implementations invent their own scheme, than the observed patterns will identify the implementation, and that's a moderate privacy risk. The other issue is of course ossification. if middle boxes "learn" that there are a few one specific patterns for that bit. The current spec specifies a random constant, which does entail a moderate ossification risk.

Specifying nothing amount to trusting the implementations. That will probably lead implementers to copy whatever Google does, in order to not stick out. That may be fine. Specifying random would work if implementations were really using good random sources, but implementations have very often failed to do that. If they use a predictable pseudo-random pattern, they can be easily fingerprinted. 

I am looking at the "interop" reason. The main risk is that an overzealous receiver barfs because the spin bit is not what it expected. So maybe we should just say that endpoints that do not enable the spin bit MAY put whatever value they see fit, and MUST accept any incoming value.

Reply:
I would be fine with what you propose here.

Reply:
OK, I just updated the text by removing the specification.

Reply:
Correction:

> this text is a PR, so it needs consensus to be added at all

That isn't the standard we've been holding to.  That will begin soon, but not yet.

As for the specific recommendation: I think that the concrete requirement is that the "spin/not" decision has to be made per-path, but the "how to not spin" decision is one best left to implementation discretion - other than perhaps recommending that the value does not leak any private information (about what is encrypted, or about what might happen on other paths for the same connection).

Reply:
I don't like unspecified. Why not just set it to zero and apply the header protection mask bit that is otherwise not applied. This means AEAD sees it as 0, and middleboxes sees it as random. Setting it to anything else is not going to fool endpoints wrt. deactivated spinning. Using header protection mask also avoids PRNG issues in packet construction. Having to store extra PRNG state is just a burden and grows state.

Reply:
To do what? I'm not trying to make it impossible to determine whether an endpoint is spinning; I'm trying to avoid a constant 0 or 1 signal when we've elsewhere tried not to do that.

Reply:
> To do what? 
Leaking that the endpoint is not spinning

But if you don't care about the endpoint finding out, why try to hide it. Maybe masking the bit requires more coordination than worthwhile - I am not too keen on this either, but alternatives are not really a lot better. Setting random per packet before AEAD would make it random on wire and would not require endpoint coordination, but it is yet another thing to deal with in header protection (tracking PRNG state to set a bit per packet). Leaving it undefined and unmasked would make the pattern arbitrarily detectable by middleboxes - fingerprinting as @huitema says - this is bad.

> I'm trying to avoid a constant 0 or 1 signal when we've elsewhere tried not to do that.

That is mostly on wire, no? TP values have lots of 0 signals by default.

Reply:
Yes, this would be on the wire.  We're talking about creating the impression that a particular value of the bit carries a particular meaning.

Reply:
I am going to mark this conversation as resolved, since the most recent commits address EKR review.